### PR TITLE
Upsert webhooks

### PR DESCRIPTION
**Pull request by kzh**
This PR adds in the option to update webhooks if they already exist under the specified name instead of creating a new one. It also changes the HMAC signature to conform to GitHub specifications.

### CODE REVIEW

File: src/main/java/com/sourcegraph/webhook/Dispatcher.java [hunk 1]
```diff
@@ -56,7 +47,7 @@ public Dispatcher(Webhook hook, EventSerializer serializer) {
 
     private static String sign(String secret, String data) {
         HmacUtils hmac = new HmacUtils(HmacAlgorithms.HMAC_SHA_256, secret);
-        return hmac.hmacHex(data);
+        return "sha256=" + hmac.hmacHex(data);
```
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/Dispatcher.java:50]: This is an unrelated change, yes?

File: src/main/java/com/sourcegraph/webhook/registry/WebhookRegistry.java [hunk 2]
```diff
@@ -78,16 +77,43 @@ public WebhookRegistry(ActiveObjects ao, ProjectService projects, RepositoryServ
...
         String scope = resolveScopeID(hook.scope);
 
-        activeObjects.executeInTransaction(() -> {
-            Map<String, Object> params = new HashMap<>();
-            params.put("NAME", hook.name);
-            params.put("SCOPE", scope);
-            params.put("ENDPOINT", hook.endpoint);
-            params.put("SECRET", hook.secret);
+        WebhookEntity[] ents = activeObjects.find(WebhookEntity.class, Query.select().where("ID = ?", hook.id));
+        if (ents.length == 1) {
```
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/registry/WebhookRegistry.java]: Would be more readable to:

```
if (ents.length == 0) {
   create(...)
} else {
   update(...)
}
```

Comment (tsenart) [reply src/main/java/com/sourcegraph/webhook/registry/WebhookRegistry.java]: Don't we need to wrap the read and write cycle in a transaction? i.e. `activeObjects.find` needs to be ran in the scope of a transaction if its results are going to be modified and written back.

**Review by tsenart — CHANGES_REQUESTED** [review]

File: src/main/java/com/sourcegraph/webhook/Dispatcher.java (hunk 1 above)
Comment (kzh) [reply src/main/java/com/sourcegraph/webhook/Dispatcher.java:50]: Yes, this is to conform to GitHub specifications. In the automation code, we are using `gh.ValidateSignature` for verifying HMAC signatures.

**Review by mrnugget — APPROVED** [review]

File: src/main/java/com/sourcegraph/webhook/registry/WebhookRegistry.java [hunk 3]
```diff
@@ -79,23 +78,26 @@ public static void register(Webhook hook) throws WebhookException {
...
+            WebhookEntity[] ents = activeObjects.find(WebhookEntity.class, Query.select().where("ID = ?", hook.id));
+            WebhookEntity ent = ents.length == 0 ? activeObjects.create(WebhookEntity.class) : ents[0];
+            ent.setName(hook.name);
+            ent.setScope(scope);
+            ent.setScope(hook.endpoint);
+            ent.setScope(hook.secret);
+            ent.save();
+
+            for (EventEntity event : ent.getEvents()) {
+                activeObjects.delete(event);
```
Comment (tsenart) [inline src/main/java/com/sourcegraph/webhook/registry/WebhookRegistry.java]: Why are we deleting events and re-creating them?

Comment (kzh) [reply src/main/java/com/sourcegraph/webhook/registry/WebhookRegistry.java]: I changed the code to remove old events and create new ones instead of destroying/recreating all.

Comment (tsenart) [reply src/main/java/com/sourcegraph/webhook/registry/WebhookRegistry.java]: I don't understand why this is necessary. Can you explain?

Comment (kzh) [reply src/main/java/com/sourcegraph/webhook/registry/WebhookRegistry.java]: This is to update (upsert) registered events for a webhook. For example: if a webhook is registered for `pr:comment:added` + `pr:comment:deleted` and we want to update it to `pr:reviewers:updated`, this logic would remove the older events and add in the desired one. 

I realize most of this logic won't be (isn't) used at all for automation. I plan to overhaul/remove it in 3.13. It is here purely for a complete REST implementation.

Comment (tsenart) [reply src/main/java/com/sourcegraph/webhook/registry/WebhookRegistry.java]: > his logic would remove the older events and add in the desired one.

So, are events stored in the database forever regardless of registered webhooks? I don't understand why there needs to be any event stored — we should be only sending it out if there's a registered webhook that matches.

> I realize most of this logic won't be (isn't) used at all for automation. I plan to overhaul/remove it in 3.13. It is here purely for a complete REST implementation.

How exactly is it not used? Please elaborate and be explicit about what work you're proposing to do in 3.13.
